build/datagen-steps/metrics.js

```diff
@@ -0,0 +1,425 @@
+import csv from "csvtojson";
+
+import siteConfigData from "../../data/config/site.js";
+import dataConfigData from "../../data/config/data.js";
+
+const dataConfig = dataConfigData.default;
+const siteConfig = siteConfigData.default;
+
+/** Return true if `n` is convertible to a number. */
+function isNumeric(n) {
+  return !Number.isNaN(Number.parseFloat(n)) && Number.isFinite(Number(n));
+}
+
+/**
+ * Transform CSV array in v1 format to JSON object.
+ *
+ * Format of v1 CSV:
+ *
+ * - Each row represents all years of data for a particular FIPS region
+ * - The first column (`id`) is always present and represents the FIPS code
+ * - Subsequent columns are created for each year of data (y_2013, y_2014, etc.)
+ *
+ * For example:
+ *
+ * ```csv
+ * id, y_2013, y_2014, y_2015;
+ * 370630001011, 17, 17, 18;
+ * 370630001012, 13, 14, 13;
+ * ```
+ *
+ * The output JSON object is the same as for the v2 importer. It is keyed by the FIPS code, and
+ * contains values in the {y_2012: value, ...} format. For example:
+ *
+ * ```json
+ *  {
+ *    "370630001011": {
+ *      "y_2013": 17,
+ *      "y_2014": 17,
+ *      "y_2015": 18
+ *  },
+ *    "370630001012": {
+ *      ...
+ *    }
+ *  }
+ * ```
+ */
+function v1CsvToJson(csvArray) {
+  const jsonOut = {};
+  for (let i = 0; i < csvArray.length; i++) {
+    jsonOut[csvArray[i].id] = {};
+    Object.keys(csvArray[i]).forEach((key) => {
+      if (key !== "id") {
+        if (isNumeric(csvArray[i][key])) {
+          jsonOut[csvArray[i].id][key] = Number(csvArray[i][key]);
+        } else {
+          jsonOut[csvArray[i].id][key] = null;
+        }
+      }
+    });
+  }
+  return jsonOut;
+}
+
+/**
+ * Transform CSV array in v2 format to JSON object.
+ *
+ * Format of v2 CSV:
+ *
+ * - Each row represents one year of data for a particular FIPS region
+ * - The columns are always `year, fips, value`
+ *
+ * For example:
+ *
+ * ```csv
+ * year, fips, value;
+ * 2013, 370630001011, 17;
+ * 2014, 370630001011, 17;
+ * 2015, 370630001011, 18;
+ * 2013, 370630001012, 13;
+ * 2014, 370630001012, 14;
+ * 2015, 370630001012, 13;
+ * ```
+ *
+ * The output JSON object is the same as for the v1 importer. It is keyed by the FIPS code, and
+ * contains values in the {y_2012: value, ...} format. For example:
+ *
+ * ```json
+ *  {
+ *    "370630001011": {
+ *      "y_2013": 17,
+ *      "y_2014": 17,
+ *      "y_2015": 18
+ *  },
+ *    "370630001012": {
+ *      ...
+ *    }
+ *  }
+ * ```
+ */
+function v2CsvToJson(csvArray) {
+  const jsonOut = {};
+  csvArray.forEach((row) => {
+    if (!("fips" in row && "year" in row && "value" in row)) {
+      throw new TypeError("Each row must have keys fips, year and value");
+    }
+    if (!row.fips || !row.year) {
+      throw new TypeError("Each row must have a valid fips and year value");
+    }
+    if (!(row.fips in jsonOut)) {
+      jsonOut[row.fips] = {};
+    }
+    jsonOut[row.fips][`y_${row.year}`] = isNumeric(row.value) ? Number(row.value) : null;
```

Comment:
Is there a reason this line uses ternary logic but lines 57-60 are an if/else? Think the earlier lines could be tweaked. Also, it'd be great to also truncate the number of decimal places here to something like 4 or 5, in which case having a utility function for both cases (something like `cleanNumber`) might make sense

Reply:
Oh, only because this difference between the two functions was carried over from what was there previously. I did not notice the discrepancy. Will fix. And roger that with truncation.